### PR TITLE
Add QNX support

### PR DESCRIPTION
**Pull request by iDings**
Support heap-profile and cpu-profile. 

Stil WIP, some issues need to be resolved, see https://github.com/gperftools/gperftools/issues/1429#issuecomment-1726837434

### CODE REVIEW

File: benchmark/unwind_bench.cc [hunk 1]
```diff
@@ -7,7 +7,11 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <stdint.h>
+#if defined(__QNXNTO__)
+#include <ucontext.h>
+#else
 #include <sys/ucontext.h>
```
Comment (alk) [inline benchmark/unwind_bench.cc:13]: I think it is worth just dropping includes of ucontext stuff here. It is conditionally included below only for platforms that we're able to use it.

File: m4/pc_from_ucontext.m4 [hunk 2]
```diff
@@ -46,6 +46,10 @@ AC_DEFUN([AC_PC_FROM_UCONTEXT],
    pc_fields="$pc_fields uc_mcontext->ss.srr0"         # OS X (ppc, ppc64 [untested])
    pc_fields="$pc_fields uc_mcontext->__ss.__srr0"     # OS X (>=10.5 [untested])
    pc_fields="$pc_fields uc_mcontext->__ss.__pc"       # OS X (arm64)
+   pc_fields="$pc_fields uc_mcontext.cpu.eip"          # QNX (7.1 x86 [untested])
+   pc_fileds="$pc_fields uc_mcontext.gpr[[ARM_REG_PC]]"  # QNX (7.1 arm [untested])
+   pc_fields="$pc_fields uc_mcontext.cpu.rip"          # QNX (7.1 x86_64)
+   pc_fields="$pc_fields uc_mcontext.cpu.elr"          # QNX (7.1 aarch64)
    pc_field_found=false
```
Comment (alk) [inline m4/pc_from_ucontext.m4:53]: thanks for this. Happy to take it now if you're able to split it to distinct commit (but do note my comment on the ticket about whether it is worth at all to try supporting cpu profiling on qnx)

File: src/base/sysinfo.cc [hunk 3]
```diff
@@ -62,6 +62,9 @@
 #include <process.h>          // for getpid() (actually, _getpid())
 #include <shlwapi.h>          // for SHGetValueA()
 #include <tlhelp32.h>         // for Module32First()
+#elif defined(__QNXNTO__)
```
Comment (alk) [inline src/base/sysinfo.cc:65]: this proc/pid/maps stuff is LGTM too.

File: src/profile-handler.cc [hunk 4]
```diff
@@ -382,7 +388,11 @@ ProfileHandler::ProfileHandler()
   // Install the signal handler.
   struct sigaction sa;
   sa.sa_sigaction = SignalHandler;
+#ifdef SA_RESTART
```
Comment (alk) [inline src/profile-handler.cc:391]: perhaps easier and cleaner is something like that instead (outside of function body)

```
#ifndef SA_RESTART
#define SA_RESTART 0
#endif
```

**Comment by iDings** [comment]
I will split this patch

1.  heap profile, which maybe fine to merge  now 
2.  stacktrace with  libunwind support, which still have some issues, and need some more work to fix it. 
4.  cpu profile, which unsure whether it's really needed on qnx

File: src/stacktrace_libunwind-inl.h [hunk 5]
```diff
@@ -45,20 +45,58 @@ extern "C" {
...
+  void operator--() {
+    int counter = reinterpret_cast<intptr_t>(perftools_pthread_getspecific(recursive_key_));
+    --counter;
+    perftools_pthread_setspecific(recursive_key_, reinterpret_cast<void*>(counter));
+  }
+
+private:
+  pthread_key_t recursive_key_;
+};
+static ThreadCounter recursive;
```
[the remark this answers is not in the document]
Comment (iDings) [reply src/stacktrace_libunwind-inl.h:98]: yeah, this one is bad, 
> My main question is it is actually necessary to support anti-recursion detection on QNX ?
 I an unsure, i will do some more analysis.

From previous comments about anti-recursion, this seems the issue from libunwind,  need to figure out whether libunwind in QNX also have the same behavior.
https://github.com/gperftools/gperftools/blob/589d416977977f32fde82cd623a7a876ab1d397c/src/stacktrace_libunwind-inl.h#L54-L61

**Comment by iDings** [comment]
Compile for QNX, still need use addition `CPPFLAGS` mentioned https://github.com/gperftools/gperftools/issues/1147#issuecomment-1004137185
```
./configure --host=aarch64-unknown-nto-qnx7.1.0 --enable-frame-pointers CPPFLAGS="-D__NO_EXT_QNX -D_SC_NPROCESSORS_ONLN=91"
```

From the compiler log, `cfree` and `mallopt` declaration is different between QNX libc and gperftools, 
```c
src/libc_override_gcc_and_weak.h:215:8: error: conflicting declaration of C function 'void cfree(void*)'
   void cfree(void* ptr) __THROW                   ALIAS(tc_cfree);
        ^~~~~
In file included from src/debugallocation.cc:44: 
target/qnx7/usr/include/malloc.h:64:12: note: previous declaration 'int cfree(void*)'
 extern int cfree(void *__ptr);

src/libc_override_gcc_and_weak.h:225:7: error: conflicting declaration of C function 'int mallopt(int, int)'
   int mallopt(int cmd, int value) __THROW         ALIAS(tc_mallopt);
target/qnx7/usr/include/malloc.h:140:12: note: previous declaration 'int mallopt(int, intptr_t)'
extern int mallopt(int __cmd, intptr_t __value);
```

Force alias seems also not a good idea, @alk any suggestions, or just it go.
```c
#ifdef __QNXNTO__
  int cfree(void* ptr) __THROW                   ALIAS(tc_cfree);
#else
  void cfree(void* ptr) __THROW                   ALIAS(tc_cfree);
#endif
  void* memalign(size_t align, size_t s) __THROW  ALIAS(tc_memalign);
  void* aligned_alloc(size_t align, size_t s) __THROW ALIAS(tc_memalign);
  void* valloc(size_t size) __THROW               ALIAS(tc_valloc);
  void* pvalloc(size_t size) __THROW              ALIAS(tc_pvalloc);
  int posix_memalign(void** r, size_t a, size_t s) __THROW
      ALIAS(tc_posix_memalign);
#ifndef __UCLIBC__
  void malloc_stats(void) __THROW                 ALIAS(tc_malloc_stats);
#endif
#ifdef __QNXNTO__
  int mallopt(int cmd, intptr_t value) __THROW         ALIAS(tc_mallopt);
#else
  int mallopt(int cmd, int value) __THROW         ALIAS(tc_mallopt);
#endif
```

**Comment by iDings** [comment]
Updates: on QNX socketpair api need link socket library, weired previous compile is passed.
And stash cpu profiler and libunwind support on an seperated branch:
cpu profiler support on https://github.com/iDings/gperftools/tree/qnx_cpu_profiler, 
libunwind support on https://github.com/iDings/gperftools/tree/qnx_libunwind

**Comment by iDings** [comment]
Have an issue:
 compile is fine, but `libc++.so` and `libsocket.so` didn' t in the NEEDED shared library,  need to add this two library in 
 LD_PRELOAD to run the application.  Previous test can work is because the test application also link this two library and have this two library in the NEEDED library and linker can link this two library. Need some more analysis to figure out why this happened.
```shell
 $eu-readelf -Wd  libtcmalloc.so.9
Dynamic segment contains 32 entries:
 Addr: 0x000000000006e7f8  Offset: 0x05e7f8  Link to section: [ 5] '.dynstr'
  Type              Value
  NEEDED            Shared library: [libunwind.so.8]
  NEEDED            Shared library: [liblzma.so.5]
  NEEDED            Shared library: [libz.so.2]
  NEEDED            Shared library: [libm.so.3]
  NEEDED            Shared library: [libc.so.5]
  NEEDED            Shared library: [libgcc_s.so.1]
  SONAME            Library soname: [libtcmalloc.so.9]
...
```

**Comment by iDings** [comment]
Updates:
1. For QNX autoconfig compile, need force config `CC`` and `CXX` to `qcc` and `q++`. otherwise the compiler didn't include `libc++.so.1` in output `libtcmalloc` library. 
```shell
./configure CC="qcc -Vgcc_ntoaarch64le" CXX="q++ -Vgcc_ntoaarch64le" --host=aarch64-unknown-nto-qnx7.1.0 --enable-frame-pointers CXXFLAGS="-D__NO_EXT_QNX -D_SC_NPROCESSORS_ONLN=91"
```
2. For socketpair, QNX need link an extra `socket` library.  Need add an sepearted m4 macro like as m4/acx_nanosleep.m4?

**Comment by ibogosavljevic** [comment]
i am curious if this is going to be merged? Can I use this branch as it is?

**Comment by iDings** [comment]
@ibogosavljevic I also unsure whether it will be merged or not,  this patch is just work, not good.  I think you can use my [qnx_port](https://github.com/iDings/gperftools/tree/qnx_port) branch directly if you want.

**Comment by ibogosavljevic** [comment]
What are the compilation instructions?

**Comment by iDings** [comment]
> ```shell
> ```shell
> ./configure CC="qcc -Vgcc_ntoaarch64le" CXX="q++ -Vgcc_ntoaarch64le" --host=aarch64-unknown-nto-qnx7.1.0 --enable-frame-pointers CXXFLAGS="-D__NO_EXT_QNX -D_SC_NPROCESSORS_ONLN=91"
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

**Comment by iDings** [comment]
> What are the compilation instructions?
for qnx aarch64, use
```
./configure CC="qcc -Vgcc_ntoaarch64le" CXX="q++ -Vgcc_ntoaarch64le" --host=aarch64-unknown-nto-qnx7.1.0 --enable-frame-pointers
```

**Comment by ibogosavljevic** [comment]
I compiled it with 
```
./autogen.sh
 ./configure CC="qcc -Vgcc_ntoaarch64le" CXX="q++ -Vgcc_ntoaarch64le" --host=aarch64-unknown-nto-qnx7.1.0 --enable-frame-pointers CXXFLAGS="-D__NO_EXT_QNX -D_SC_NPROCESSORS_ONLN=91" --enable-minimal --enable-cpu-profiler
``` 

But in the .libs folder I don't see libprofiler.so.

**Comment by iDings** [comment]
> Updates: on QNX socketpair api need link socket library, weired previous compile is passed.
> 
> And stash cpu profiler and libunwind support on an seperated branch: cpu profiler support on https://github.com/iDings/gperftools/tree/qnx_cpu_profiler, libunwind support on https://github.com/iDings/gperftools/tree/qnx_libunwind

@ibogosavljevic for cpu profile, you also need cherry-pick the change from https://github.com/iDings/gperftools/tree/qnx_cpu_profiler,  because QNX don't support SIGPROF,  the cpu profiler use wall-time,  see https://github.com/gperftools/gperftools/issues/1429#issuecomment-1728142330

**Comment by alk** [comment]
Thanks a lot. I took first commit. Second I replaced with my own (which I think is better; let me know if you disagree). I am able to compile this with simple ./configure --host=x86_64-pc-nto-qnx7.1.0

I.e. no need to define odd things, hopefully. (But I am unable to test qnx binaries)

**Comment by iDings** [comment]
> Thanks a lot. I took first commit. Second I replaced with my own (which I think is better; let me know if you disagree). I am able to compile this with simple ./configure --host=x86_64-pc-nto-qnx7.1.0

@alk thanks for your review.  
you already had QNX SDP?  If yes, QNX SDP have a [mkqnximage](https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.utilities/topic/m/mkqnximage.html) which can use to create a qemu qnx virtual machine.  
This link https://doc.rust-lang.org/rustc/platform-support/nto-qnx.html#rust-std-library-test-suite have some guideline about how to use `mkqnximage`

**Pull request closed without merging**